### ISSUE TITLE
Vault export download overwrites files when filename set

### ISSUE THREAD

**Issue by daethnir** [taers232c/GAMADV-XTD3#65]
Note: this may be a failure to craft the correct command line, rather than a fault in gam. If so then the solution is an update to the wiki with more examples, rather than a gam source change.

Trying to download the files associated with vault exports. For example, when an email export is created, here are the relevant bits you'd see from `gam print vaulteports`

```
matterId: MATTERID-guid-guid-guid-blahblahblah
matterName: user_defined_mattername
id: exportly-XXXXXXXX-guid-guid-guid-blahblahblah
name: user_defined_name
cloudStorageSink.files.0.bucketName: BUCKETNM-GUID-GUID-GUID-blahblahblah
cloudStorageSink.files.0.objectName: MATTERID-guid-guid-guid-blahblahblah/exportly-XXXXXXXX-guid-guid-guid-blahblahblah/user_defined_name-1.zip
cloudStorageSink.files.1.bucketName: BUCKETNM-GUID-GUID-GUID-blahblahblah
cloudStorageSink.files.1.objectName: MATTERID-guid-guid-guid-blahblahblah/exportly-XXXXXXXX-guid-guid-guid-blahblahblah/user_defined_name-metadata.xml
cloudStorageSink.files.2.bucketName: BUCKETNM-GUID-GUID-GUID-blahblahblah
cloudStorageSink.files.2.objectName: MATTERID-guid-guid-guid-blahblahblah/exportly-XXXXXXXX-guid-guid-guid-blahblahblah/user_defined_name-results-count.csv

```

there are three files, <mattername>-1.zip, <mattername>-metadata.xml, and <mattername>results-count.csv. We'd like to download them all.

You can do so via

```gam download vaultexport id:exportly-XXXXXXXX-guid-guid-guid-blahblahblah matter id:MATTERID-guid-guid-guid-blahblahblah noextract```

And the resulting files are named

```MATTERID-guid-guid-guid-blahblahblah-exportly-XXXXXXXX-guid-guid-guid-blahblahblah-user_defined_name-1.zip
MATTERID-guid-guid-guid-blahblahblah-exportly-XXXXXXXX-guid-guid-guid-blahblahblah-user_defined_name-results-count.csv
...
```

I'd love to be able to download the files one by one using targetname to give them useful names, presumably user_defined_name-1.zip, user_defined_name-results-count.csv.

Or alternatively to pull them down in one go while replacing everything before the extension (i.e. matterid + id + name) with the value from `targetname`

Currently, when using `targetname` it sets the target file name, however since it's downloading 3 files it ends up overwriting the same file, thus leaving you with the contents of last one downloaded.

The relevant code snippet from `gam.py`:

```
22868   for s_file in export['cloudStorageSink']['files']:
22869     j += 1
22870     bucket = s_file['bucketName']
22871     s_object = s_file['objectName']
22872     filename = os.path.join(targetFolder, s_object.replace('/', '-'))
22873     if zipToStdout and not ZIP_EXTENSION_PATTERN.match(filename):
22874       continue
22875     if targetName:
22876       filename = targetName
```

We see there on line 22876 it simply sets the name, does not handle any modifications to maintain the suffix.

Options:

* document better command line that lets you download a file at a time and use `targetname`
* update code above to pre-pend `targetname` to original `objectName` value
* update code above to grok the `objectName` format, strip all of the name above the extension, and replace with `targetname`

**Comment by taers232c** [comment]
Call me on Hangouts or send a Meet invitation.

Ross

**Comment by taers232c** [comment]
Try this:
```
    if targetName:
      _, s_objectFilename = s_object.rsplit('/', 1)
      _, s_objectExtension = s_object.rsplit('.', 1)
      if targetName.find('#') == -1:
        filename = targetName+'.'+s_objectExtension
      else:
        filename = targetName.replace('#objectname#', s_object).targetName.replace('#filename#', s_objectFilename).replace('#extension#', s_objectExtension)
```

**Comment by taers232c** [comment]
Handle multiple -n.zip files
```
    if targetName:
      _, s_objectFilename = s_object.rsplit('/', 1)
      if s_objectFilename.endswith('.zip'):
        _, s_objectExtension = s_objectFilename.rsplit('-', 1)
        s_objectExtension = '-'+s_objectExtension
      else:
        _, s_objectExtension = s_objectFilename.rsplit('.', 1)
        s_objectExtension = '.'+s_objectExtension
      if targetName.find('#') == -1:
        filename = targetName+s_objectExtension
      else:
        filename = targetName.replace('#objectname#', s_object).targetName.replace('#filename#', s_objectFilename).replace('#extension#', s_objectExtension)
```

**Comment by taers232c** [comment]
Third time's a charm:
```
COUNT_ZIP_EXTENSION_PATTERN = re.compile(r'^.*(-\d+\.zip)$', re.IGNORECASE)

# gam download vaultexport|export <ExportItem> matter <MatterItem> [targetfolder <FilePath>] [targetname <FileName>] [noverify] [noextract] [ziptostdout]                                                                                                                                                                                    
# gam download vaultexport|export <MatterItem> <ExportItem> [targetfolder <FilePath>] [targetname <FileName>] [noverify] [noextract] [ziptostdout]                                                                                                                                                                                           
def doDownloadVaultExport():
....
    if targetName:
      _, s_objectFilename = s_object.rsplit('/', 1)
      mg = COUNT_ZIP_EXTENSION_PATTERN.match(s_objectFilename)
      if mg:
        s_objectExtension = mg.group(1)
      else:
        _, s_objectExtension = s_objectFilename.rsplit('.', 1)
        s_objectExtension = '.'+s_objectExtension
      if targetName.find('#') == -1:
        filename = targetName+s_objectExtension
      else:
        filename = targetName.replace('#objectname#', s_object).targetName.replace('#filename#', s_objectFilename).replace('#extension#', s_objectExtension)
```

**Comment by taers232c** [comment]
OK, maybe the fourth time.

```
    if targetName:
      _, s_objectFilename = s_object.rsplit('/', 1)
      mg = COUNT_ZIP_EXTENSION_PATTERN.match(s_objectFilename)
      if mg:
        s_objectExtension = mg.group(1)
      else:
        _, s_objectExtension = s_objectFilename.rsplit('.', 1)
        s_objectExtension = '.'+s_objectExtension
      if targetName.find('#') == -1:
        filename = targetName+s_objectExtension
      else:
        filename = targetName.replace('#objectname#', s_object).replace('#filename#', s_objectFilename).replace('#extension#', s_objectExtension)
      filename = os.path.join(targetFolder, filename.replace('/', '-'))
```

**Comment by taers232c** [comment]
Fixed 5.03.40

**Comment by taers232c** [comment]
Did 5.03.40 fix your issue?
